Content Localization Part

jptissot opened the pull request:
Fixes  #2081 

Creating this PR for feedback. 
Content linking via LocalizationSet seems to work.
There are probably lots of edge cases I missed. 
The LocalizationPart UI need some love.

agriffard commented:
I started from a blank site, enable the feature and added the LocalizationPart on a Page.

Here are a few remarks after testing the Content Localization features:

First, I tested without any cultures so the System culture is the default one in the general Settings.
When I create a new page, an empty Content Culture list is displayed.
Then, I added the English culture.
Should we display it (or a warning) if no culture is set or only one?
if French is added and set as default, English is still displayed first, because it is sorted alphabetically I guess.

In Edit, the list is displayed as disabled. 
Shouldn't we display a label instead or hide it, as it will appear in the list below?

The culture code and the English culture name are displayed.
What should description we display?
We could display it as Bootstrap list group, with a special class for the currently selected culture and may be in a collapsible panel.

In the Content items list:
Should we display the culture in each content summary that has a LocalizationPart?
If the feature is enabled, should we display a Culture Filter?

agriffard commented:
Have you planned to add a Culture Layer rule to be able to only display some widgets depending on the current culture?

As an example, I tried to add a Widget stereotype to the Menu content type.
We need to think about it because when people will want to display a different navigation based on the culture, they will probably do like this (and find a way to remove the menu if it has been added manually in the Layout view).

It seems that an error occurs when you add a Widget.
The first time, this Exception appears:
```
NullReferenceException: Object reference not set to an instance of an object.
OrchardCore.ContentLocalization.Records.LocalizedContentItemIndexProvider+<>c.<Describe>b__0_0(ContentItem contentItem) in LocalizedContentItemIndex.cs, line 26

NullReferenceException: Object reference not set to an instance of an object.
                    if (string.IsNullOrEmpty(localizationPart.LocalizationSet))
```

and then, it is impossible to add a Widget anymore in a Layer.

jptissot commented:
I have been super busy lately but I plan on working on this sometime this week.

jptissot commented:
Antoine, Thanks for your comments! Did a bit more work on this tonight.

> I started from a blank site, enable the feature and added the LocalizationPart on a Page.
> 
> Here are a few remarks after testing the Content Localization features:
> 
> First, I tested without any cultures so the System culture is the default one in the general Settings.
> When I create a new page, an empty Content Culture list is displayed.

Fixed, now displaying System culture with link to create cultures.

> Then, I added the English culture.
> Should we display it (or a warning) if no culture is set or only one?

Fixed I think
> In the Content items list:
> Should we display the culture in each content summary that has a LocalizationPart?
Added a driver for the SummaryAdmin. Needs some placement/ styling.

I also fixed the exception with the index when attaching a Widget

jptissot commented:
We should probably also group all localized ContentItems in `Admin/Contents/ContentItems` view with a link to each localization.

agriffard commented:
Thank you @jptissot .
There aren't errors anymore.

The display of the summary admin and the list in details can still be improved.

The `Deleted` label that appears next to a culture that has been removed is a bit disturbing.

The link to the culture settings can be handy but once the cultures are set, it won't be used anymore so is it really necessary?.

As explained, have you considered to add a Culture Layer rule that you can use in the layers?

jptissot commented:
@agriffard Thanks!

Yeah, I am mostly focused on functionality right now so expect ugly UI ;)

What do you recommend we do about deleted cultures that have localized items? That is what the Deleted label is meant to represent. I agree with you though it makes it seem like the ContentItem was deleted.  I am not sure it is relevant to display this information though.

> The link to the culture settings can be handy but once the cultures are set, it won't be used anymore so is it really necessary?.

I can remove that link. 

Do you find the Dropdown that allows selecting Culture of the Current Item useful ?

How would you style the list of links to other cultures ?

I did not consider Culture Layer rule yet. I was going to look into the CulturePicker soon. Do you have any recommendations other than ading a Layer rule ?

jptissot commented:
@agriffard I made some UX improvements and a couple bugfixes (if a localized content item was deleted for example, the index behaved strangly) Let me know what you think ! Onto the CulturePicker next!

agriffard commented:
Do we really want to allow to change the culture in Edit by providing a list of the cultures not already localized?

If we add some css (for the accordion), it will be better to have it in a distinct file, not in the view.

About the CulturePicker, will it be a Widget than you can add in the layer you want?

Skrypt commented:
I pushed my changes on a different branch please review them.

Skrypt commented:
Will make a change here. Instead of using InvariantCulture we should maybe use the System Default Culture.

Skrypt commented:
After thinking about this we should never have a content item set to InvariantCulture since our frontend website is always displayed on a specific culture. The culture fallback mechanism that we use with the website culture settings should be replicated.

Content Item culture > culture site setting > system (OS) culture setting.

So, if someone has content items that have no LocalizationPart attached it should always be assigned to a default culture using this order.

Skrypt commented:
Ok I pushed an update to https://github.com/OrchardCMS/OrchardCore/tree/skrypt/content-localization
Maybe we should now merge it on the PR here if you guys approve the changes.

jptissot commented:
@agriffard I made some improvements after yesterday's demo. Let me know what you think.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs:
```diff
@@ -7,28 +7,24 @@
-        public async Task<IActionResult> Localize(string contentItemId, string targetCulture)
-        {
+
+        public async Task<IActionResult> Localize(string contentItemId, string targetCulture) {
```
Please update all your changes to follow the coding style (curly braces on new lines)
There is an editorsettings file in the solution that VS should pick up automatically.

jptissot replied:
Ok will do, I was using vscode on linux today.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.ContentLocalization/Assets.json:
```diff
@@ -0,0 +1,9 @@
+    ],
+    "output": "wwwroot/Styles/contentLocalization.css"
+  }
+]
```
Don't forget to add empty lines at the end of text files (I did it for you)

sebastienros commented:
TODO: check in O1 if we should reassign the Owner with the current user.

sebastienros commented:
`ICultureManager.GetDefaultCulture`
-> If Settings.DefaultCulture is defined, return it, or return system default culture.

Same projects as ILocalClock and LocalClock.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs:
```diff
@@ -0,0 +1,71 @@
+                _notifier.Information(T["Successfully created localized version of the content."]);
+                return RedirectToAction("Edit", "Admin", new { area = "OrchardCore.Contents", contentItemId = newContent.ContentItemId });
+            }
+            catch (InvalidOperationException)
```
I checked `LocalizeAsync` and it throws a `NotSupportedException`. I think `InvalidOperationException` is the one to use.

sebastienros commented on src/OrchardCore/OrchardCore.ContentLocalization.Core/DefaultContentLocalizationManager.cs:
```diff
@@ -0,0 +1,87 @@
+            var cloned = await _contentManager.CloneAsync(content);
+            var clonedPart = cloned.As<LocalizationPart>();
+            clonedPart.Culture = targetCulture;
+            //TODO: Remove next line. This is nessesary because of a bug with "Flushing" the cloned data to the parts
```
There is an Alter method that flushes automatically the changes

jptissot replied:
Can I use it for Clone ? I would like Clone to do it for all parts automatically so Indexes get updated.

sebastienros replied:
Alter is calling Apply and just doing a lookup for the part, so this code is fine, keep it this way.

sebastienros commented on src/OrchardCore/OrchardCore.ContentLocalization.Core/Records/Migrations.cs:
```diff
@@ -0,0 +1,36 @@
+            );
+
+            SchemaBuilder.AlterTable(nameof(LocalizedContentItemIndex), table => table
+              .CreateIndex("IDX_LocalizationPartIndex_ContentItemId", "ContentItemId")
```
I think we need an index for every single column

sebastienros commented on src/OrchardCore.Modules/OrchardCore.ContentLocalization/Indexing/LocalizationPartIndexHandler.cs:
```diff
@@ -0,0 +1,22 @@
+
+            foreach (var key in context.Keys)
+            {
+                context.DocumentIndex.Set(key, part.LocalizationSet, options);
```
They both need to be stored. You can force it in the options object. Because I assume we should be able to read any of these values if we filter by the other one.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.ContentLocalization/Manifest.cs:
```diff
@@ -0,0 +1,11 @@
+    Author = "The Orchard Team",
+    Website = "http://orchardproject.net",
+    Version = "1.0.0",
+    Description = "The Content Localization module provides a part that allows you to easily localize content",
```
```suggestion
    Description = "Provides a part that allows to localize content items.",
```

sebastienros commented on src/OrchardCore/OrchardCore.ContentLocalization.Core/Handlers/LocalizationPartHandler.cs:
```diff
@@ -0,0 +1,84 @@
+        private async Task<string> GetDefaultCulture()
+        {
+            // TODO: change this to the correct default culture if we choose to do so
+            string defaultCulture = CultureInfo.InstalledUICulture.Name;
```
@Skrypt can you update the recommended action here while this is fresh in your mind?

Skrypt replied:
In the views where we display the CultureInfo.DisplayName or CultureInfo.Name since we support InvariantCulture we should do something like this every time.
```
                            @if (!String.IsNullOrWhiteSpace(culture.Name))
                            {
                                @culture.Name <text> | </text>
                            }
                            @if (!String.IsNullOrWhiteSpace(culture.DisplayName))
                            {
                                @culture.DisplayName
                            }
                            else
                            {
                                @culture.NativeName
                            }
```

Skrypt replied:
The change is not related to that line specifically. I think we said that we should just move this entire method in a ILocalCulture service instead. The code seems correct. The issue is within the Views themselves when we display a content item that is set as InvariantCulture.

sebastienros commented:
@jptissot I published the first review. There are other points I will probably add later but there is some things to do already.

I tried to localize a blog post and it worked well. But it also made me realize that for List we should translate the list itself because the two lists might have different items. But maybe some lists are not like blogs and users will want to provide a localized version of each blog post, but in this case they'll have to filter which posts to render so they won't be able to use our default List rendering.

I also see that the ContentField is not handled during localization. This can be done separately but we should not forget. Maybe when we implement invariant culture for field. Using the handler we need to search for each localized version of a referenced content item and change the reference to this one.

hishamco commented:
Note from DNN
> You can’t change the default site language after enabling content localization, so make sure you have selected the correct default language before enabling the content localization.

I vote for this suggestion because it will save us from a lot of issues and scenarios if the user will be able to change the default culture after enabling content localization module

hishamco commented:
Few suggestion:

- Localize the content when it's saved with the current default culture, no need for two steps, save and localize

- Change the way of localizing the content, it's too wired from UI/UX perspective, I prefer shows flags or cultures in badge (I can share a picture for what it looks like my local version)

- Manage Urls, if this is done with AutoRoute it will be good otherwise, we need sort of `UrlRequestCultureProvider`

- Group the content items by culture, this will be great for site editors, they can see which content is not translated yet .. again with flags will be obvious

hishamco commented:
Is there a way to push some commits here? or I need to fork @jptissot repo and send a PR?

hishamco commented:
![Localization1](https://user-images.githubusercontent.com/3237266/57195064-d735ab00-6f56-11e9-9b37-994db085192e.png)

![Localization2](https://user-images.githubusercontent.com/3237266/57195069-dbfa5f00-6f56-11e9-9e67-deed2a32e15b.png)

sebastienros commented:
Maybe DNN can't change their default culture because they use a default value like `null` for it. We talked about it during the meeting. We don't have this issue with the current design.

> save us from a lot of issues and scenarios if the user will be able to change the default culture after enabling content localization module

Start by sharing "one" issue and we can talk about it ;)

sebastienros commented:
The design with the flags looks nice, an maybe we can switch to the drop down version only after a specific number of cultures. Usually sites should handle 2-3 languages and that would be better. I assume the flags are always displayed and grayed out when the content is not yet localized to this culture?

hishamco commented:
> I assume the flags are always displayed and grayed out when the content is not yet localized to this culture?

Exactly

> Start by sharing "one" issue and we can talk about it ;)

Me as a site administrator I expect when I add a content after enable content localization that the added content is localized to the default culture not null or empty string, I am not sure I need an extra step and choose my default culture?!!

sebastienros commented:
What you'd expect is what we do. So there are no issues and we can change the default culture if we need without issues.

jptissot commented:
I don't mind flags but what about when a country has two languages (en-ca, fr-ca) ?

jptissot commented:
How about I fix the issues Sébastien mentioned in his review and we can have a follow up ticket for ux?

hishamco commented:
> what about when a country has two languages (en-ca, fr-ca) ?

Good point .. we should display the region which is Canada, but we can display the language on hover

BTW all the UI changes is already done but I don't know how can I participate on this PR

hishamco commented:
> what about when a country has two languages (en-ca, fr-ca) ?

Good point .. we should display the region which is Canada, but we can display the language on hover

BTW all the UI changes is already done but I don't know how can I participate on this PR

sebastienros commented on src/OrchardCore/OrchardCore.ContentLocalization.Core/DefaultContentLocalizationManager.cs:
```diff
@@ -0,0 +1,87 @@
+            var cloned = await _contentManager.CloneAsync(content);
+            var clonedPart = cloned.As<LocalizationPart>();
+            clonedPart.Culture = targetCulture;
+            //TODO: Remove next line. This is nessesary because of a bug with "Flushing" the cloned data to the parts
```
```suggestion

```

sebastienros commented on src/OrchardCore/OrchardCore.ContentLocalization.Core/Handlers/LocalizationPartHandler.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace OrchardCore.ContentLocalization.Handlers
+{
+    public class LocalizationPartHandler : ContentPartHandler<LocalizationPart>
```
I don't understand why you have a handler for this, I would assume assigning the culture is done after the content is cloned and it's localized. Because at that point you have a `LocalizationPart` you can just assign the default culture there, if no specific culture was specified.

jptissot replied:
I think the handler was for when a content is not localized but you attach a localization part to it.
I will try to see if It breaks.

sebastienros commented:
I am done with the changes. Would appreciate that you all try it and ensure I didn't break much of it ;)

Regarding flags, I was googling about it and found this site: http://www.flagsarenotlanguages.com/blog/
It has lots of posts looking at websites and how they represent locales. My overall conclusion is that flags are hard at conveying locales. But read them by yourself.

jptissot commented:
@sebastienros I tested the following use cases and they work:

- New blog tenant, attach LocalizationPart to BlogPost, Save, Localize.
- New BlogPost when no default clulture is assigned (using system culture)
- Change default culture then try to create / edit currently localized Content Items

I made a couple changes, cleaned the code a little bit.

Is there a clean way to add a modal dialog saying that you will loose your changes when clicking on the Edit / Create localized content buttons ?

> TODO: check in O1 if we should reassign the Owner with the current user. (on the clone event)
Did you still want to do this ?

hishamco commented:
> Is there a clean way to add a modal dialog saying that you will loose your changes when clicking on the Edit / Create localized content buttons ?

Check the bootstrap modal that I created in the admin

iboshkov commented:
Was the localization picker ever created ?

jptissot commented:
@iboshkov  see https://github.com/OrchardCMS/OrchardCore/pull/3819 https://github.com/OrchardCMS/OrchardCore/pull/3813 

Maybe these videos will help: 
- https://www.youtube.com/watch?v=wldI6SOkaqo&t=530s
- https://www.youtube.com/watch?v=QBkuJeaKrGQ&t=815s

Writing documentation is my next priority once these get merged.

iboshkov commented:
Ah, thanks. Still poking through Orchard Core, will add some docs myself if I can.
